qiskit/pulse/library/symbolic_pulses.py

```diff
@@ -100,43 +98,55 @@ class LamdifiedExpression:
     For the next time it will just return the cached callbacks for speed up.
     """
 
-    def __init__(self, common_params: List[str], attribute: str):
+    def __init__(self, attribute: str):
         """Create new descriptor.
 
         Args:
-            common_params: Parameters that the expression always takes regardless
-                of pulse instances.
             attribute: Name of attribute of :class:`.SymbolicPulse` that returns
                 the target expression to evaluate.
         """
-        self.common_params = common_params
         self.attribute = attribute
         self.lambda_funcs = dict()
 
     def __get__(self, instance, owner) -> Callable:
         expr = getattr(instance, self.attribute, None)
+        if expr is None:
+            raise PulseError(f"'{self.attribute}' of '{instance.pulse_type}' is not assigned.")
         key = hash(expr)
         if key not in self.lambda_funcs:
-            if expr is None:
-                raise PulseError(f"'{self.attribute}' of '{instance.pulse_type}' is not assigned.")
             self.__set__(instance, expr)
 
         return self.lambda_funcs[key]
 
     def __set__(self, instance, value):
         key = hash(value)
         if key not in self.lambda_funcs:
-            import sympy as sym
-
-            params = [sym.Symbol(p) for p in self.common_params + instance._param_names]
-            self.lambda_funcs[key] = sym.lambdify(params, value)
+            if value.free_symbols == instance.parameters.keys():
+                raise PulseError(
+                    "Symbolic pulse parameter and expression's free symbols don't match. "
+                    "Cannot generate samples for this symbolic pulse."
+                )
+            params = sorted(value.free_symbols, key=lambda s: s.name)
+
+            if _optional.HAS_SYMENGINE:
+                # Symengine lambdify requires array-like
+                try:
+                    lamb = sym.lambdify(params, [value])
+                except RuntimeError:
+                    # When sym.I is used in the expression, i.e. complex envelope.
+                    lamb = sym.lambdify(params, [value], real=False)
+                func = np.vectorize(lamb)
```

Comment:
You are correct. I used vectorize for the sake of code simplicity, but I was not aware of such performance regression. With vectorize the `.get_waveform` call for `GaussianSquare(1600, 0.1, 64, 1344)` is very slow as you mentioned.

![image](https://user-images.githubusercontent.com/39517270/172567414-a0c7ffc7-c3f7-4e83-93a4-aaac31b2e5aa.png)

But this overhead is significantly reduced with extended array in https://github.com/Qiskit/qiskit-terra/pull/7821/commits/bef913e989fa7731604228a8bc1106a2991ed382 (> x10 improvement).

![image](https://user-images.githubusercontent.com/39517270/172567691-23f14767-a7ab-41b5-aaf0-641dba9f3105.png)

I guess this is memory inefficient but maybe okey for standard pulses with short duration.

Reply:
Minor, but perhaps we could ask the symengine developers if numpy broadcasting could ever be supported for a lamdified expression, so that a mix of array and scalar arguments could work.

Reply:
I've written the issue for this (in bullet3)
https://github.com/symengine/symengine.py/issues/400

> That's something we can fix here.

Sounds like they plan to support this. The rest of remarkable performance bottleneck comes from array generation, so this code will be much faster if they support broadcasting in C++ level.